### PR TITLE
[#5518] feat(iceberg): add http header and remote addr information to Iceberg Event 

### PR DESCRIPTION
**Pull request by FANNG1**
### What changes were proposed in this pull request?
1. add HTTP header to event listener context
2. add remote IP addr to event listener context

### Why are the changes needed?

Fix: #5518 

### Does this PR introduce _any_ user-facing change?
no

### How was this patch tested?
setup a local app and check IP addr and HTTP header was added to context

### CODE REVIEW

File: iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/service/IcebergRestUtils.java [hunk 1]
```diff
@@ -122,6 +126,19 @@ public static NameIdentifier getGravitinoNameIdentifier(
     return NameIdentifier.of(catalogNS.toArray(String[]::new));
   }
 
+  public static Map<String, String> getHttpHeaders(HttpServletRequest httpServletRequest) {
+    Map<String, String> headers = new HashMap<>();
+    Enumeration<String> headerNames = httpServletRequest.getHeaderNames();
+    while (headerNames.hasMoreElements()) {
+      String headerName = headerNames.nextElement();
+      String headerValue = httpServletRequest.getHeader(headerName);
+      if (headerValue != null) {
```
Comment (puchengy) [inline iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/service/IcebergRestUtils.java:135]: Curious is there a chance that headerValue can be null here?

File: iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/service/rest/IcebergTableOperations.java [hunk 2]
```diff
@@ -82,7 +82,7 @@ public class IcebergTableOperations {
   private ObjectMapper icebergObjectMapper;
   private IcebergTableOperationDispatcher tableOperationDispatcher;
 
-  @Context private HttpServletRequest httpRequest;
+  @Inject private Provider<HttpServletRequest> httpRequest;
```
Comment (puchengy) [inline iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/service/rest/IcebergTableOperations.java:85]: Curious why is this needed?

**Review by puchengy — APPROVED** [review]
lgtm, thank you

File: iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/service/IcebergRestUtils.java (hunk 1 above)
Comment (FANNG1) [reply iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/service/IcebergRestUtils.java:135]: According to the API, the headerValue will be null if `headerName` is not in HTTP request headers. In theory the headerValue will not be null in current implement, but it's safe to add the check here.

File: iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/service/rest/IcebergTableOperations.java (hunk 2 above)
Comment (FANNG1) [reply iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/service/rest/IcebergTableOperations.java:85]: `TestTableOperations` failed because httpRequest is null ,   I try to fix the test failure by using `Inject`, but still not fixed.

File: iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/service/IcebergRestUtils.java (hunk 1 above)
Comment (puchengy) [reply iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/service/IcebergRestUtils.java:135]: Thanks!

**Comment by FANNG1** [comment]
@jerryshao @puchengy it's ready to review now, please help to review again.

**Comment by FANNG1** [comment]
After this PR is merged, you could add Http header information to Gravitino server event too, and then make audit frame work to generate audit log for both Gravitino server and Gravitino IcebergREST server.

**Comment by puchengy** [comment]
nit, the title of the PR should be expanded to include ip address (and possibly remote hostname if my suggestion is implemented).

**Comment by puchengy** [comment]
@FANNG1 Thank you, LGTM, left some nits.

File: iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/service/IcebergRequestContext.java [hunk 3]
```diff
@@ -19,16 +19,23 @@
...
 import lombok.Getter;
+import org.apache.gravitino.utils.PrincipalUtils;
 
 /** The general request context information for Iceberg REST operations. */
 public class IcebergRequestContext {
-  @Getter private final HttpServletRequest httpRequest;
+
   @Getter private final String catalogName;
+  @Getter private final String userName;
+  @Getter private final String remoteIp;
```
[the remark this answers is not in the document]
Comment (FANNG1) [reply iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/service/IcebergRequestContext.java:32]: remoteHostName and remoteIP contains same information, I perfer just keep one,  do you think remoteHostName is more useful than remoteIP?

File: iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/service/rest/IcebergTableOperations.java [hunk 4]
```diff
@@ -269,6 +269,10 @@ public Response reportTableMetrics(
     return IcebergRestUtils.noContent();
   }
 
+  HttpServletRequest httpServletRequest() {
```
[the remark this answers is not in the document]
Comment (FANNG1) [reply iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/service/rest/IcebergTableOperations.java:274]: update the comment

File: iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/service/IcebergRequestContext.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jerryshao) [reply iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/service/IcebergRequestContext.java]: I think it should be made clear whether the return value is IP address or hostname, or others, it should not be mixed conceptually.

Besides, do we need a remote port as well?

[the remark this answers is not in the document]
Comment (jerryshao) [reply iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/service/IcebergRequestContext.java]: This class now becomes a public class that will be leveraged by users. You should move it to the `listener/api/event` package, also:

1. don't use lombok to generate the method, clearly define them manually and add javadoc.
2. make the class stable and make all the event compatible.

File: iceberg/iceberg-rest-server/src/test/java/org/apache/gravitino/iceberg/service/rest/IcebergRestTestUtil.java [hunk 5]
```diff
@@ -119,4 +124,11 @@ protected void configure() {
     }
     return resourceConfig;
   }
+
+  static HttpServletRequest createMockHttpRequest() {
+    HttpServletRequest mockRequest = mock(HttpServletRequest.class);
+    when(mockRequest.getRemoteAddr()).thenReturn("127.0.0.1");
+    when(mockRequest.getHeaderNames()).thenReturn(Collections.emptyEnumeration());
+    return mockRequest;
+  }
```
Comment (jerryshao) [inline iceberg/iceberg-rest-server/src/test/java/org/apache/gravitino/iceberg/service/rest/IcebergRestTestUtil.java:133]: Can you please add several UTs to cover your change for `Context` and various `events`.

File: iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/service/IcebergRequestContext.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (FANNG1) [reply iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/service/IcebergRequestContext.java]: updated with `remoteHostName`, because it's more user friendly, if it has some performance decrease for resolving hostname, let's change to `remoteIP`

[the remark this answers is not in the document]
Comment (FANNG1) [reply iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/service/IcebergRequestContext.java]: As to the remote port , seems not necessary,  databricks and HDFS audit doesn't contain the remote port too.

[the remark this answers is not in the document]
Comment (FANNG1) [reply iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/service/IcebergRequestContext.java]: @jerryshao  what do you mean `make the class stable and make all the event compatible.` ?

File: iceberg/iceberg-rest-server/src/test/java/org/apache/gravitino/iceberg/service/rest/IcebergRestTestUtil.java (hunk 5 above)
Comment (FANNG1) [reply iceberg/iceberg-rest-server/src/test/java/org/apache/gravitino/iceberg/service/rest/IcebergRestTestUtil.java:136]: Updated UT

File: iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/service/IcebergRequestContext.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (FANNG1) [reply iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/service/IcebergRequestContext.java]: keep old httpServeletRequest in context, so old listener could use the HTTP request in the event.

**Review by jerryshao — APPROVED** [review]
